rdflib/plugins/parsers/patch.py

```diff
@@ -0,0 +1,150 @@
+                              See `.W3CNTriplesParser.parse`
+        """
+        assert sink.store.context_aware, (
+            "RDFPatchParser must be given" " a context aware store."
+        )
+        # type error: Incompatible types in assignment (expression has type "ConjunctiveGraph", base class "W3CNTriplesParser" defined the type as "Union[DummySink, NTGraphSink]")
+        self.sink: Dataset = Dataset(store=sink.store)  # type: ignore[assignment]
+        self.skolemize = skolemize
+
+        source = inputsource.getCharacterStream()
+        if not source:
+            source = inputsource.getByteStream()
+            source = getreader("utf-8")(source)
+
+        if not hasattr(source, "read"):
+            raise ParseError("Item to parse must be a file-like object.")
+
+        self.file = source
+        self.buffer = ""
+        while True:
+            self.line = __line = self.readline()
+            if self.line is None:
+                break
+            try:
+                self.parsepatch(bnode_context)
+            except ParseError as msg:
+                raise ParseError("Invalid line (%s):\n%r" % (msg, __line))
+
+        if self.skolemize:
+            return self.sink
+        else:
+            self.sink = self.sink.de_skolemize()
+            return self.sink  # Dataset is skolemized as part of adding/removing triples
+            # , so de skolemize before returning. NB this is broken by the parent class
+            # (ConjunctiveGraph) which re-skolemizes the dataset.
+
+    def parsepatch(self, bnode_context: Optional[_BNodeContextType] = None) -> None:
+        self.eat(r_wspace)
+        #  From spec: "No comments should be included (comments start # and run to end
+        #  of line)."
+        if (not self.line) or self.line.startswith("#"):
+            return  # The line is empty or a comment
+
+        # if header, transaction, skip
+        operation = self.operation()
+        self.eat(r_wspace)
+
+        if operation == Operation.AddTripleOrQuad:
+            self.add_triple_or_quad()
+        elif operation == Operation.DeleteTripleOrQuad:
+            self.delete_triple_or_quad()
+        elif operation == Operation.AddPrefix:
+            self.add_prefix()
+        elif operation == Operation.DeletePrefix:
+            self.delete_prefix()
+
+    def add_triple_or_quad(self):
+        self.sink.parse(data=self.line, format="nquads", skolemize=True)
+
+    def delete_triple_or_quad(self):
+        removal_ds = Dataset()
+        removal_ds.parse(data=self.line, format="nquads", skolemize=True)
+        triple_or_quad = next(iter(removal_ds))
+        self.sink.remove(triple_or_quad)
+
+    def add_prefix(self):
+        # Extract prefix and URI from the line
+        prefix, ns, _ = self.line.replace('"', "").replace("'", "").split(" ")
+        ns_stripped = ns.strip("<>")
+        self.sink.bind(prefix, ns_stripped)
```

Comment:
This adds the new prefix to the new local Dataset instance created on line 70, not the original dataset passed in. Is that an issue? Do the new prefixes appear on the original dataset's namespace manager because they share a store?

Reply:
> Do the new prefixes appear on the original dataset's namespace manager because they share a store?
Yes - I've tried binding the prefixes to the original instance and the behaviour is the same.
When the Dataset is instantiated and parse is called, the method creates a ConjunctiveGraph instance, which in turn creates a Graph instance which parse is called on. This is probably why the nquads parser (which I've followed) creates another Dataset instance. Would it be best to remove the creation of a ConjunctiveGraph instance when Dataset.parse() is called in a separate PR, which should remove the need to create a local Dataset instance in the parser? (+ probably the nquads parser too).

Reply:
Discussed w/ Nick & Ashley not making further changes to this PR, but future tasks to:
- discuss Dataset / Graph model / interfaces (add, parse etc.)
- simplify/standardise parse/serialize interfaces generally